### ISSUE TITLE
Double played note

### ISSUE THREAD

**Issue by oddtime** [hydrogen-music/hydrogen#1512]
playing in pattern mode, if a note is placed on tick 0 it's triggered two times (I hear two attacks)

**Comment by cme** [comment]
Interesting... can you share a reproducible song example? Is it with playback in pattern mode? That would imply that at least one of the notes entered in the queue is played at the wrong time! 

But also: the current ADSR implementation (which I'm fixing...) shares ADSR state between all notes on the same instrument. So if a note is followed immediately by another note on the same instrument, no matter how soft the velocity is, it will restart the attack for the previous note, so there's a possibility that might be what you're encountering ?

**Comment by oddtime** [comment]
> Is it with playback in pattern mode?

yes

> can you share a reproducible song example?

just enter a note in tick 0 and launch the playback

**Comment by theGreatWhiteShark** [comment]
@oddtime which commit are you working on? There have been quite huge changes in the audio engine lately.

**Comment by oddtime** [comment]
> which commit are you working on?

I have just pulled from master

**Comment by cme** [comment]
Confirmed, I hear the same thing.

**Comment by cme** [comment]
A quick bisect, it sounds like this came in with 0fecc5af76ee0567ff3e346ac1d26835fa25bd2e . Sorry @theGreatWhiteShark ...

**Comment by theGreatWhiteShark** [comment]
> A quick bisect, it sounds like this came in with https://github.com/hydrogen-music/hydrogen/commit/0fecc5af76ee0567ff3e346ac1d26835fa25bd2e . Sorry @theGreatWhiteShark ...

Well, since I had to rewrite half of the audio engine for fixing the tempo marker support I expected to miss at least some glitches. Let's just hope we catch all major ones before the proper release. So, thanks a lot @oddtime for reporting!

I'll have a look in 1-2 days. (A couple of hours ago I was lucky enough to bring my JACK server in a state neither it nor ALSA are able to start properly and Hydrogen is not able to do any audio processing. I don't want to spoil this setup yet :) )

**Comment by theGreatWhiteShark** [comment]
Hmm. I can't reproduce it on 6d2386464dd435df822dc66dd6898ec622858835. I created a new song, inserted a single note and started playback in both pattern and song mode. Either way and regardless of picking JACK, ALSA, or PortAudio driver I just hear and see a single note.

There are two commented debug messages regarding such problems in the code: 
1. https://github.com/hydrogen-music/hydrogen/blob/master/src/core/AudioEngine/AudioEngine.cpp#L2267 for notes encountered by the AudioEngine
2. https://github.com/hydrogen-music/hydrogen/blob/master/src/core/AudioEngine/AudioEngine.cpp#L1268 for handing the notes over to the Sampler

Do you see two notes instead of one being picked in 1.?

**Comment by oddtime** [comment]
Yes two notes from the debug messages uncommenting the line
https://github.com/hydrogen-music/hydrogen/blob/master/src/core/AudioEngine/AudioEngine.cpp#L2267 


```
(D) AudioEngine::updateNoteQueue getDoubleTick(): 374.491, getFrames(): 172032, nnTick: 383, [Note], instrument_id: 1, specific_compo_id: -1, position: 0, m_nNoteStart: 0, m_fUsedTickSize: nan, velocity: 0.8, pan: 0, length: -1, pitch: 0, key: 0, octave: 0, [[ADSR] attack: 0, decay: 0, sustain: 1, release: 999, state: 0, ticks: 0, value: 0, release_value: 0], lead_lag: 0, cut_off: 1, resonance: 0, humanize_delay: 0, key: 0, bpfb_l: 0, bpfb_r: 0, lpfb_l: 0, lpfb_r: 0, pattern_idx: 0, midi_msg: -1, note_off: 0, just_recorded: 0, probability: 1, instrument: Stick, layers_selected: [component: 0, selected layer: -1, sample position: 0]                               
(D) AudioEngine::updateNoteQueue getDoubleTick(): 375.606, getFrames(): 172544, nnTick: 384, [Note], instrument_id: 1, specific_compo_id: -1, position: 0, m_nNoteStart: 0, m_fUsedTickSize: nan, velocity: 0.8, pan: 0, length: -1, pitch: 0, key: 0, octave: 0, [[ADSR] attack: 0, decay: 0, sustain: 1, release: 999, state: 0, ticks: 0, value: 0, release_value: 0], lead_lag: 0, cut_off: 1, resonance: 0, humanize_delay: 0, key: 0, bpfb_l: 0, bpfb_r: 0, lpfb_l: 0, lpfb_r: 0, pattern_idx: 0, midi_msg: -1, note_off: 0, just_recorded: 0, probability: 1, instrument: Stick, layers_selected: [component: 0, selected layer: -1, sample position: 0]                               
(W) Sampler::renderNote sample position out of bounds. The layer has been resized during note play?                                   
(W) Sampler::renderNote sample position out of bounds. The layer has been resized during note play? 
```

It happens after two repetitions in Pattern mode

**Comment by theGreatWhiteShark** [comment]
> (D) AudioEngine::updateNoteQueue getDoubleTick(): 374.491, getFrames(): 172032, [...]                    
(D) AudioEngine::updateNoteQueue getDoubleTick(): 375.606, getFrames(): 172544, [...]                               

Ah, thanks! I see. I had my buffer size set to 1024. But for smaller values I can reproduce it too.

**Comment by theGreatWhiteShark** [comment]
Hey @oddtime,

Could you check whether #1524 fixes this issue?

**Comment by oddtime** [comment]
> Could you check whether https://github.com/hydrogen-music/hydrogen/pull/1524 fixes this issue?

Thanks, I have tested it and it is working fine

**Comment by theGreatWhiteShark** [comment]
Closed with #1524
